chebi-update/src/main/java/org/reactome/release/chebiupdate/ChebiUpdater.java

```diff
@@ -0,0 +1,487 @@
+			}
+		}
+	}
+
+	/**
+	 * Update a molecule's identifier.
+	 * 
+	 * @param molecule
+	 * @param chebiID
+	 * @param moleculeIdentifier
+	 * @param prefix
+	 * @throws InvalidAttributeException
+	 * @throws InvalidAttributeValueException
+	 * @throws Exception
+	 */
+	private void updateMoleculeIdentifier(GKInstance molecule, String chebiID, String moleculeIdentifier, String prefix) throws InvalidAttributeException, InvalidAttributeValueException, Exception
+	{
+		if (!chebiID.equals(moleculeIdentifier))
+		{
+			molecule.setAttributeValue("identifier", chebiID);
+			this.identifierSB.append(prefix).append(" Old Identifier: ").append(moleculeIdentifier).append(" ; ").append("New Identifier: ").append(chebiID).append("\n");
+			if (!testMode)
+			{
+				adaptor.updateInstanceAttribute(molecule, "identifier");
+			}
+		}
+	}
+
+	/**
+	 * Updates ReferenceEntities that refer to ReferenceMolecules. This method will
+	 * ensure that the "name" array of a ReferenceEntity associated with
+	 * <code>molecule</code> has <code>chebiName</code> as the <em>first</em> name
+	 * in its list of names.
+	 * 
+	 * @param molecule - a ReferenceMolecule. ReferenceEntities that reference this ReferenceMolecule will be updated.
+	 * @param chebiName - the name from ChEBI. Should be the first name in the "name" array for any updated ReferenceEntity.
+	 * @throws Exception
+	 * @throws InvalidAttributeException
+	 * @throws InvalidAttributeValueException
+	 */
+	private void updateReferenceEntities(GKInstance molecule, String chebiName) throws Exception, InvalidAttributeException, InvalidAttributeValueException
+	{
+		@SuppressWarnings("unchecked")
+		Collection<GKInstance> refEntities = (Collection<GKInstance>) molecule.getReferers("referenceEntity");
+
+		for (GKInstance refEntity : refEntities)
+		{
+			@SuppressWarnings("unchecked")
+			LinkedList<String> names = new LinkedList<String>(refEntity.getAttributeValuesList("name"));
+			// Now we must ensure that the name from the ChEBI molecule is the FIRST name in
+			// the referenceEntity's list of names.
+			if (!names.isEmpty())
+			{
+				// name[0] does not match - but before adding chebName, ensure that it's not
+				// already somewhere else in the array.
+				if (!names.get(0).toLowerCase().equals(chebiName.toLowerCase()))
+				{
+					int i = 0;
+					boolean nameFound = false;
+					while (!nameFound && i < names.size())
+					{
+						String name = names.get(i);
+						// We found the chebi name so now we swap.
+						if (name.toLowerCase().equals(chebiName.toLowerCase()))
+						{
+							nameFound = true;
+							// Remove the ChEBI name at position i
+							names.remove(i);
+							// Add the ChEBI name back at position 0.
+							names.add(0, chebiName);
+						}
+						i++;
+					}
+					// If we went through the whole array and didn't find the ChEBI name,
+					// we must add it, at the begining of the list of names.
+					if (!nameFound)
+					{
+						names.add(0, chebiName);
+					}
```

Comment:
is this block new?

Reply:
This is to check that if the name from the ChEBI service hasn't been found in the list of existing names by the time all existing names have been checked, then add the name to the beginning of the list.

Similar functionality exists in the Perl code but I think this code is easier to understand (even if it has more lines)